compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1070,6 +1070,40 @@ fn assemble_candidates_from_impls<'cx, 'tcx>(
                         | ty::Infer(..)
                         | ty::Error(_) => false,
                     }
+                } else if lang_items.async_destruct_trait() == Some(trait_ref.def_id) {
+                    match self_ty.kind() {
+                        ty::Bool
+                        | ty::Char
+                        | ty::Int(_)
+                        | ty::Uint(_)
+                        | ty::Float(_)
+                        | ty::Adt(..)
+                        | ty::Str
```

Comment:
So we implement it for all of these, too, but then the impl is a noop? Could we just skip them like we do for `Copy` types in the regular `Drop` glue?

Reply:
No, `async_drop_in_place` must return async destructor future for any type, thus there should be a constructor of async destructor shim for any type.

Reply:
oh 🤦 makes sense, we can invoke `drop_in_place` on copy types after all.. Please document this here, that just like everything implements the hypothetical `Destruct` trait, everything needs to implement `AsyncDestruct`. Even `Copy` types.